lib/models/finamp_models.dart

```diff
@@ -270,60 +341,109 @@ class FinampSettings {
+      required this.baseDirectory}) {
+    assert(baseDirectory.needsPath == (relativePath != null));
+    assert(baseDirectory == DownloadLocationType.migrated ||
+        (legacyUseHumanReadableNames == null && legacyDeletable == null));
+    assert(baseDirectory != DownloadLocationType.migrated ||
+        (legacyUseHumanReadableNames != null && legacyDeletable != null));
+  }
 
   /// Human-readable name for the path (shown in settings)
   @HiveField(0)
   String name;
 
   /// The path. We store this as a string since it's easier to put into Hive.
   @HiveField(1)
-  String path;
+  String? relativePath;
 
   /// If true, store songs using their actual names instead of Jellyfin item IDs.
+  @Deprecated("This is here for migration.  Use useHumanReadableNames instead.")
   @HiveField(2)
-  bool useHumanReadableNames;
+  bool? legacyUseHumanReadableNames;
+
+  bool get useHumanReadableNames => baseDirectory.useHumanReadableNames;
+  bool get needsPermission => baseDirectory.needsPermission;
 
   /// If true, the user can delete this storage location. It's a bit of a hack,
   /// but the only undeletable location is the internal storage dir, so we can
   /// use this value to get the internal song dir.
   @HiveField(3)
-  bool deletable;
+  @Deprecated("This is here for migration.  Use baseDirectory instead.")
+  bool? legacyDeletable;
 
   /// Unique ID for the DownloadLocation. If this DownloadLocation was created
   /// before 0.6, it will be "0", very temporarily until it is changed on
   /// startup.
   @HiveField(4, defaultValue: "0")
   String id;
 
+  /// Base directory of DownloadLocation.  Used to calculate currentPath and
+  /// to determine directory attributes.
+  @HiveField(5, defaultValue: DownloadLocationType.migrated)
+  DownloadLocationType baseDirectory;
+
+  String? _currentPath;
+
+  /// The current path to the location, updated during app startup
+  String get currentPath => _currentPath!;
+
+  /// Update curentPath to the latest value.  Run fro every downloadLocation
+  /// every time the app starts up.
+  Future<void> updateCurrentPath() async {
+    if (baseDirectory == DownloadLocationType.migrated) {
+      if (!legacyDeletable!) {
+        baseDirectory = DownloadLocationType.internalDocuments;
+        relativePath = null;
+        name = "Legacy Internal Storage";
+      } else if (!legacyUseHumanReadableNames!) {
+        baseDirectory = DownloadLocationType.external;
+      } else {
+        baseDirectory = DownloadLocationType.custom;
+      }
+      legacyDeletable = null;
+      legacyUseHumanReadableNames = null;
+    }
+    switch (baseDirectory) {
+      case DownloadLocationType.internalDocuments:
+        _currentPath = (await getApplicationDocumentsDirectory()).path;
+      case DownloadLocationType.internalSupport:
+        _currentPath = (await getApplicationSupportDirectory()).path;
+      case DownloadLocationType.external:
+        // TODO add more logic? Are these paths guaranteed stable?
```

Comment:
Theoretically, we could enumerate valid external directories and search for a song or a breadcrumb file or something, because there's only a few.  This could be useful if either the OS or the user moves the downloads to a differently named directory/storage device.  But that seems complicated, the existing code didn't have any provisions for these paths moving, and I don't have a personal use-case that would tell me how useful this feature may or may not be, so I never set that up.

Reply:
yeah I don't think it's a good idea or necessary either. If users move the files, things are allowed to break :)